subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java

```diff
@@ -295,6 +299,13 @@ private static void configureBuiltInTest(Project project, DefaultJvmSoftwareComp
         project.getTasks().named(JavaBasePlugin.CHECK_TASK_NAME, task -> task.dependsOn(testSuite));
     }
 
+    private static void configureDiagnostics(Project project, JvmSoftwareComponentInternal component) {
+        // TODO: Deprecate this convention?
```

Comment:
This seems like a very useful convention to me, most of the time you'll have questions about the dependencies on the compilation classpath.  We should re-implement it with the Provider API though instead of convention mapping.

Reply:
The issue with this is that this is the only diagnostics task of its kind to have a convention like this. All other diagnostics tasks have some default that lists all options or requires some user-specified value. 

I don't think sensible defaults are a bad thing, but this convention is way too targeted. We need a more generic approach for configuring conventions for ALL diagnostics tasks

Reply:
I don't understand what you mean.  Each report is a different task type.  I think this `DependencyInsightReport` report, which is meant to determine why a dependency was resolved as it was, is most likely to be invoked about the `compileClasspath`.  Different conventions for different report types seems unavoidable.

A bigger roadblock is that the `configuration` property is not a `Property`.  We could add a new property `Property<Configuration> getConfigurationToInspect()`, update the backing field used by the current property, and go through a deprecation cycle, but that's definitely annoying for a very small gain.

It's the baked-in `HelpTasksPlugin` that creates this report, so even if we had the `Property` we couldn't configure the convention in the plugin that adds the task, as we typically try to do.  Seems like this convention will always remain a little unique.

Reply:
I've removed this comment for now since I don't think we really need to take care of this now.

> Different conventions for different report types seems unavoidable.

This would be true if all reports had different conventions, though this is not the case. While `dependencyInsight` for example has a convention on the compile classpath, the `dependencies` report does not. Instead it just prints the results for all configurations. Granted, I will say this is not really helpful and I usually want to specify a specific configuration manually. 

I think calling `dependencyInsight` on compile classpath or runtime classpath are equally valuable and useful. It seems a little weird to default to compile here. 

Whatever we do, we should be consistent. Either all of these configuration/dependency/component insights have defaults or they don't. Right now we're only setting a convention on one kind of these report tasks. We'll be adding more of these reports as we develop the component interface, so I think this is a problem we can solve in the future.